[WK2] Reduce moves/copies when decoding through the general ArgumentCoder specializations

zdobersek opened the pull request:
#### 1a7b36e70573fcf071a53c98058be7284154f61c
<pre>
[WK2] Reduce moves/copies when decoding through the general ArgumentCoder specializations
<a href="https://bugs.webkit.org/show_bug.cgi?id=248963">https://bugs.webkit.org/show_bug.cgi?id=248963</a>

Reviewed by Kimmo Kinnunen.

Improve decoding efficiency in the general ArgumentCoder specializations
by avoiding the stream extraction operator when decoding from a given
decoder object, and by utilizing in-place construction of objects
wrapped in std::optional&lt;&gt;.

Using the decoder&apos;s decode&lt;T&gt;() method to retrieve an object from the
decoder is beneficial to first default-constructing a std::optional&lt;&gt;
object and then targeting it with the stream extraction operator
invocation since it avoids one move or copy operation. This is adopted
through all the ArgumentCoder specializations in WebKit&apos;s ArgumentCoders
header and implementation file.

Where already possible, in-place construction of objects wrapped in
std::optional&lt;&gt; is done through the use of std::make_optional(). This
also avoids additional move or copy operations.

The reduction in those operations is reflected in the
decoding-of-move-counter unit tests, specifically in the lowering of
move operations for different types. A unit test covering std::variant&lt;&gt;
is also added.

Another unit test covering Expected&lt;ValueType, ErrorType&gt; types will be
added separately, along with move-semantics support for those types on
the encoding side.

* Source/WebKit/Platform/IPC/ArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;CString&gt;::decode):
(IPC::ArgumentCoder&lt;String&gt;::decode):
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
(IPC::ArgumentCoder&lt;OptionSet&lt;T&gt;&gt;::decode):
(IPC::ArgumentCoder&lt;std::optional&lt;T&gt;&gt;::decode):
(IPC::ArgumentCoder&lt;Box&lt;T&gt;&gt;::decode):
* Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: <a href="https://commits.webkit.org/257621@main">https://commits.webkit.org/257621@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4eea55bbc1320b57022febcee90247ed9d8ec8c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99364 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8562 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32484 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108747 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168997 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103360 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9124 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85881 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91852 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106674 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105121 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6899 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90440 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33876 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88718 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21787 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76752 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2432 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23304 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2336 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45706 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5241 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8496 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42779 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4203 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/Platform/IPC/ArgumentCoders.h:
```diff
@@ -257,17 +252,15 @@ template<typename T, typename U> struct ArgumentCoder<std::pair<T, U>> {
     {
-        std::optional<T> first;
-        decoder >> first;
+        auto first = decoder.template decode<T>();
```
An aside, not at all necessary for this patch but I'm mentioning in order make you aware of the thought:
This kind of pattern should be reformulated as
```
auto first = decoder.template decode<T>();
auto second = decoder.template decode<U>();
if (UNLIKELY(!decoder.isValid())
    return std::nullopt;
return std::make_optional<std::pair<T, U>>(WTFMove(*first), WTFMove(*second));
```
E.g. to reduce the checks while optimising for the common no-error case .

zdobersek committed:
[WK2] Reduce moves/copies when decoding through the general ArgumentCoder specializations
https://bugs.webkit.org/show_bug.cgi?id=248963

Reviewed by Kimmo Kinnunen.

Improve decoding efficiency in the general ArgumentCoder specializations
by avoiding the stream extraction operator when decoding from a given
decoder object, and by utilizing in-place construction of objects
wrapped in std::optional<>.

Using the decoder's decode<T>() method to retrieve an object from the
decoder is beneficial to first default-constructing a std::optional<>
object and then targeting it with the stream extraction operator
invocation since it avoids one move or copy operation. This is adopted
through all the ArgumentCoder specializations in WebKit's ArgumentCoders
header and implementation file.

Where already possible, in-place construction of objects wrapped in
std::optional<> is done through the use of std::make_optional(). This
also avoids additional move or copy operations.

The reduction in those operations is reflected in the
decoding-of-move-counter unit tests, specifically in the lowering of
move operations for different types. A unit test covering std::variant<>
is also added.

Another unit test covering Expected<ValueType, ErrorType> types will be
added separately, along with move-semantics support for those types on
the encoding side.

* Source/WebKit/Platform/IPC/ArgumentCoders.cpp:
(IPC::ArgumentCoder<CString>::decode):
(IPC::ArgumentCoder<String>::decode):
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
(IPC::ArgumentCoder<OptionSet<T>>::decode):
(IPC::ArgumentCoder<std::optional<T>>::decode):
(IPC::ArgumentCoder<Box<T>>::decode):
* Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: https://commits.webkit.org/257621@main
